doc/config.rst
.. _configuration:

=============
Configuration
=============

Configuration of Services
=========================

Configuration File Format
-------------------------
   Libra uses the `Oslo configuration library <https://wiki.openstack.org/wiki/Oslo/Config>`_
   so its format is similar to other OpenStack programs.

DEFAULT Section
^^^^^^^^^^^^^^^

   The ``[DEFAULT]`` section contains generic options common to the various
   Libra utilities (worker, mgm, etc).

   .. code-block:: ini

      [DEFAULT]
      daemon = true
      user = libra
      group = libra
      verbose = false
      debug = false
      billing_enable = false
      notification_driver = []
      default_notification_level = INFO
      default_publisher_id = None
      host = localhost
      kombu_ssl_version =
      kombu_ssl_keyfile =
      kombu_ssl_certfile =
      kombu_ssl_ca_certs =
      rabbit_use_ssl = false
      rabbit_userid = guest
      rabbit_password = guest
      rabbit_host = localhost
      rabbit_port = 5672
      rabbit_hosts = []
      rabbit_virtual_host = / 
      rabbit_retry_interval = 1
      rabbit_retry_backoff = 2
      rabbit_max_retries = 0
      rabbit_ha_queues = false
      control_exchange = openstack
      amqp_durable_queues = false

   Options supported in this section:

   .. option:: daemon

      Run as a daemon. Default is 'true'.

   .. option:: user

      Specifies the user for the process when in daemon mode. Default is the
      current user.

   .. option:: group

      Specifies the group for the process when run in daemon mode.

   .. option:: verbose

      Prints more verbose output. Sets logging level to INFO from WARNING

   .. option:: debug

      Prints debug output. Sets logging level to DEBUG from WARNING

   .. option:: billing_enable

      Enables the sending of billing information to a rabbitMQ host. It sends
      create and delete loadbalancer messages as well as exists and usage 
      messages on a periodic, configurable basis. See admin_api config.

   .. option:: notification_driver

      Driver or drivers to handle sending notifications for metering / billing.
      For instance, the openstack rpc driver is 
      openstack.common.notifier.rpc_notifier.

   .. option:: default_notification_level

      Default notification level for outgoing notifications

   .. option:: default_publisher_id

      Default publisher_id for outgoing notifications

   .. option:: host

      Default host name to use in notifications. Will use default_publisher_id
      or gethostname() if not set.

   .. option:: host

      Default host name to use in notifications. Will use default_publisher_id
      or gethostname() if not set.

   .. option:: kombu_ssl_version

      SSL version to use (valid only if SSL enabled). valid values are TLSv1,
      SSLv23 and SSLv3. SSLv2 may be available on some distributions

   .. option:: kombu_ssl_keyfile

      SSL key file (valid only if SSL enabled)

   .. option:: kombu_ssl_certfile

      SSL cert file (valid only if SSL enabled)

   .. option:: kombu_ssl_ca_certs

      SSL certification authority file (valid only if SSL enabled)

   .. option:: rabbit_use_ssl

      Connect over SSL for RabbitMQ

   .. option:: rabbit_userid

      The RabbitMQ userid

   .. option:: rabbit_password

      The RabbitMQ password

   .. option:: rabbit_host

      The RabbitMQ broker address where a single node is used

   .. option:: rabbit_port

      The RabbitMQ broker port where a single node is used

   .. option:: rabbit_hosts

      RabbitMQ HA cluster host:port pairs

   .. option:: rabbit_virtual_host

      The RabbitMQ virtual host

   .. option:: rabbit_retry_interval

      How frequently to retry connecting with RabbitMQ

   .. option:: rabbit_retry_backoff

      How long to backoff for between retries when connecting to RabbitMQ

   .. option:: rabbit_max_retries

      Maximum retries with trying to connect to RabbitMQ (the default of 0
      implies an infinite retry count)

   .. option:: rabbit_ha_queues

      Use H/A queues in RabbitMQ (x-ha-policy: all). You need to wipe RabbitMQ
      database when changing this option.

   .. option:: control_exchange

      AMQP exchange to connect to if using RabbitMQ or Qpid

   .. option:: amqp_durable_queues

      Use durable queues in amqp.

Gearman Section
^^^^^^^^^^^^^^^

   The ``[gearman]`` section contains options specific to connecting to
   a Gearman job server. All of the Libra utilities will read this section
   since each connects to Gearman.

   In order to support SSL connections, it is required that all three SSL
   related options be supplied. Also, the user owning the process must be
   able to read all SSL files.

   .. code-block:: ini

      [gearman]
      servers = 10.0.0.1:4730, 10.0.0.2:4730
      poll = 1
      ssl_ca = /etc/ssl/gearman.ca
      ssl_cert = /etc/ssl/gearman.cert
      ssl_key = /etc/ssl/gearman.key

   Options supported in this section:

   .. option:: keepalive

      Enable TCP KEEPALIVE pings. Default is 'false'.

   .. option:: keepcnt

      Max KEEPALIVE probes to send before killing connection.

   .. option:: keepidle

      Seconds of idle time before sending KEEPALIVE probes.

   .. option:: keepintvl

      Seconds between TCP KEEPALIVE probes.

   .. option:: poll

      Gearman worker polling timeout. Default is 1.

   .. option:: reconnect_sleep

      Seconds to sleep between job server reconnects. Default is 60.

   .. option:: servers

      Comma-separated list of Gearman job servers and port in HOST:PORT format.

   .. option:: ssl_ca

      Gearman SSL certificate authority.

   .. option:: ssl_cert

      Gearman SSL certificate.

   .. option:: ssl_key

      Gearman SSL key.


doc/api/config.rst
API Configuration
=================

These options are specific to the API in addition to the
:doc:`common options </config>`.

Configuration File
------------------

   The ``[api]`` section is specific to the libra_api utility.  Below is an
   example:

   .. code-block:: ini

      [api]
      db_sections=mysql1
      gearman=127.0.0.1:4730
      keystone_module=keystoneclient.middleware.auth_token:AuthProtocol
      swift_basepath=lbaaslogs
      swift_endpoint=https://host.com:443/v1/
      ssl_certfile=/opt/certfile.crt
      ssl_keyfile=/opt/keyfile.key

      [mysql1]
      host=localhost
      port=3306
      username=root
      password=
      schema=lbaas
      ssl_cert=/opt/mysql_cert.crt
      ssl_key=/opt/mysql_key.key
      ssl_ca=/opt/mysql_ca.ca

   In addition to this any options that are specific to the given keystone
   module should be stored in the ``[keystone]`` section.

Command Line Options
--------------------
   .. program:: libra_api

   .. option:: --host <IP ADDRESS>

      The IP address to bind the frontend to, default is 0.0.0.0

   .. option:: --port <PORT NUMBER>

      The port number to listen on, default is 443

   .. option:: --disable_keystone

      Do not use keystone authentication, for testing purposes only

   .. option:: --db_secions <SECTIONNAME>

      Config file sections that describe the MySQL servers.  This option can
      be specified multiple times for Galera or NDB clusters.

   .. option:: --gearman <HOST:POST>

      Used to specify the Gearman job server hostname and port. This option
      can be used multiple times to specify multiple job servers.

   .. option:: --gearman_keepalive

      Use TCP KEEPALIVE to the Gearman job server. Not supported on all
      systems.

   .. option:: --gearman_keepcnt <COUNT>

      Maximum number of TCP KEEPALIVE probes to send before killing the
      connection to the Gearman job server.

   .. option:: --gearman_keepidle <SECONDS>

      Seconds of idle time on the Gearman job server connection before
      sending TCP KEEPALIVE probes.

   .. option:: --gearman_keepintvl <SECONDS>

      Seconds between TCP KEEPALIVE probes.

   .. option:: --gearman_ssl_ca <PATH>

      The path for the Gearman SSL Certificate Authority

   .. option:: --gearman_ssl_cert <PATH>

      The path for the Gearman SSL certificate

   .. option:: --gearman_ssl_key <PATH>

      The path for the Gearman SSL key

   .. option:: --keystone_module <MODULE:CLASS>

      A colon separated module and class to use as the keystone authentication
      module.  The class should be compatible with keystone's AuthProtocol
      class.

   .. option:: --swift_basepath <CONTAINER>

      The default container to be used for customer log uploads.

   .. option:: --swift_endpoint <URL>

      The default endpoint for swift.  The user's tenant ID will automatically
      be appended to this unless overridden at the log archive request.

   .. option:: --ssl_certfile <PATH>

      The path for the SSL certificate file to be used for frontend of the API
      server

   .. option:: --ssl_keyfile <PATH>

      The path for the SSL key file to be used for the frontend of the API
      server

   .. option:: --ip_filters <FILTERS>

      A mask of IP addresses to filter for backend nodes in the form
      xxx.xxx.xxx.xxx/yy

      Any backend node IP address supplied which falls outside these filters
      will result in an error for the create or node add functions.
      This option can be specified multiple times.


debian/changelog
libra (4.0-2libra1) precise; urgency=high

  * Update dependencies
  * Split into multiple packages

 -- Endre Karlson <endre.karlson@hp.com>  Mon, 03 Sep 2013 13:58:18 +0200

libra (4.0-1libra1) precise; urgency=high

  * Pin dependancy on python-gearman

 -- Endre Karlson <endre.karlson@hp.com>  Mon, 02 Sep 2013 17:59:18 +0200

libra (1.0-1) precise; urgency=low

  * Initial release.

 -- Andrew Hutchings <andrew@linuxjedi.co.uk>  Thu, 16 Nov 2012 14:26:59 +0000


libra/common/options.py
# Copyright 2012 Hewlett-Packard Development Company, L.P.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations

import os
import os.path

from oslo.config import cfg

CONF = cfg.CONF

common_cli_opts = [
    cfg.BoolOpt('daemon',
                default=True,
                help='Run as a daemon'),
    cfg.StrOpt('group',
               help='Group to use for daemon mode'),
    cfg.StrOpt('user',
               help='User to use for daemon mode'),
    cfg.BoolOpt('billing_enable',
                default=False,
                help='Enable or disable MnB notifictions')
]

gearman_opts = [
    cfg.BoolOpt('keepalive',
                default=False,
                help='Enable TCP KEEPALIVE pings'),
    cfg.IntOpt('keepcnt',
               metavar='COUNT',
               help='Max KEEPALIVE probes to send before killing connection'),
    cfg.IntOpt('keepidle',
               metavar='SECONDS',
               help='Seconds of idle time before sending KEEPALIVE probes'),
    cfg.IntOpt('keepintvl',
               metavar='SECONDS',
               help='Seconds between TCP KEEPALIVE probes'),
    cfg.IntOpt('poll',
               default=1,
               metavar='SECONDS',
               help='Gearman worker polling timeout'),
    cfg.IntOpt('reconnect_sleep',
               default=60,
               metavar='SECONDS',
               help='Seconds to sleep between job server reconnects'),
    cfg.ListOpt('servers',
                default=['localhost:4730'],
                metavar='HOST:PORT,...',
                help='List of Gearman job servers'),
    cfg.StrOpt('ssl_ca',
               metavar='FILE',
               help='Gearman SSL certificate authority'),
    cfg.StrOpt('ssl_cert',
               metavar='FILE',
               help='Gearman SSL certificate'),
    cfg.StrOpt('ssl_key',
               metavar='FILE',
               help='Gearman SSL key'),
]


def add_common_opts():
    CONF.register_opts(gearman_opts, group='gearman')
    CONF.register_cli_opts(common_cli_opts)


def check_gearman_ssl_files():
    """
    If using Gearman with SSL, validate that the SSL files exist and
    are readable by the user. It's a common problem that connections to
    Gearman will silently fail because these files cannot be read due to
    the private key being readable only by the file owner.
    """
    if 'gearman' not in CONF:
        return
    for key in ['ssl_ca', 'ssl_cert', 'ssl_key']:
        if key in CONF['gearman']:
            fname = CONF['gearman'][key]
            if fname is None:
                continue
            if not os.path.exists(fname):
                raise Exception("Gearman SSL file %s does not exist" % fname)
            if not os.access(fname, os.R_OK):
                raise Exception("Unable to read Gearman SSL file %s" % fname)
